packages/web-microkernel/src/RealmConnection.js

```diff
@@ -0,0 +1,400 @@
+import * as _ from './lodash-es/lodash.js'
+
+// Polyfill the behavior of addEventListener(type, listener, { once: true }))
+// because for an unknown reason the above isn't working in Chrome 72.
+const addEventListenerOnce = (target, type, listener) => {
+  const onceListener = (...args) => {
+    const result = listener(...args)
+    target.removeEventListener(type, onceListener)
+    return result
+  }
+
+  target.addEventListener(type, onceListener)
+}
+
+const firstEvent = (target, type) =>
+  new Promise(resolve => {
+    addEventListenerOnce(target, type, resolve)
+  })
+
+const cutoff = 25
+
+export const inspect = value => {
+  const stringified = JSON.stringify(value)
+
+  return stringified.length > cutoff
+    ? `${stringified.slice(0, cutoff)}...`
+    : stringified
+}
+
+// Create an unforgeable key.
+const Key = () => {
+  const array = new Uint32Array(1)
+  window.crypto.getRandomValues(array)
+  return array[0].toString(36)
+}
+
+const decoders = {}
+const types = []
+
+// The types are defined in the same order they are tested.
+
+types.push(
+  {
+    name: `boolean`,
+    test: _.isBoolean,
+    encoder: (context, boolean) => boolean,
+    decoder: (context, code) => code
+  },
+
+  {
+    name: `null`,
+    test: _.isNull,
+    encoder: () => null,
+    decoder: () => null
+  },
+
+  {
+    name: `number`,
+    test: _.isNumber,
+    encoder: (context, number) => number,
+    decoder: (context, code) => code
+  },
+
+  {
+    name: `string`,
+    test: _.isString,
+    encoder: (context, string) => string,
+    decoder: (context, code) => code
+  },
+
+  {
+    name: `undefined`,
+    test: _.isUndefined,
+    encoder: () => undefined,
+    decoder: () => undefined
+  },
+
+  {
+    name: `array`,
+    test: Array.isArray,
+    encoder: (context, array) =>
+      array.map(value => encodeToDictionary(context, value)),
+
+    decoder: (context, codes) =>
+      codes.map(code => decodeFromDictionary(context, code))
+  }
+)
+
+// error
+
+const isError = value => value instanceof Error
+
+const encodeError = (context, error) => ({
```

Comment:
would be nice to have resulting types annotated for example in jsdoc, otherwise it's quite difficult to image results of encodings

Reply:
The encoding is an implementation detail, not part of the API, and could change over time so I don't think it's a good idea to put it into documentation.